Add basic security module tests

korthout opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

Adds some basic tests to the security module that were still missing. In particular:
- Adds coverage of the `AuthorizationChecker`
- Adds coverage of the `MappingRuleMatcher`

It does not add coverage of the `ArgumentUtil` as these methods do not require unit testing.

This PR also re-organizes some security module code. In particular:
- Moves `AuthorizationScope` from `protocol-impl` to `security-protocol`
- Reorders security-core dependencies for keeping a better overview
- Removes some unnecessary mockito usage to test behavior rather than implementation

**Additional notes**

- Anonymous users are never authorized
- Authorized users are authorized if they have access to the resource by id or as wildcard

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [x] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #31766

korthout committed:
refactor: move test to security-protocol module

This test covers the AuthorizationScope wrapper methods. It only covers
the AuthorizationScope class and nothing else. So it makes sense to
place it in the same package in the same module.

korthout committed:
fix: adjust test to java-8

This module is java 8 only, so we cannot language features from later
versions.

korthout committed:
deps: reorder jcip-annotations in security-core

This is scoped to compile for production code. It fits better somewhere
between the production code dependencies and the test scope
dependencies.

korthout committed:
deps: reorder security-core internal dependencies

By placing the internal dependencies (io.camunda) first, it becomes
easier to get an overview of the dependencies.

We now have:
- internal dependencies
- third-party production code dependencies
- third-party compile dependencies
- third-party test dependencies

korthout committed:
test: check isAuthorized edge cases

This adds a unit test for AuthorizationChecker that verifies the correct
functioning of the isAuthorized method.

It uses parameterized tests for scenarios with anonymous users and
authenticated users, clients and mapping rules, where it considers
whether they can access a resource. Cases cover all reasonable
combinations of authentication, authorization, and resource to access.
An expected enum is used to easily clarify the expected outcome: Access
allowed, Access denied. Some scenarios overlap, but that's fine as this
test is cheap to maintain.

The AuthorizationChecker depends AuthorizationReader for which a fake
implementation is provided. This is done rather than using Mockito to
mock it, as using a fake leads to tests that are less dependent on the
implementation of AuthorizationChecker and how it interacts with
AuthorizationReader, but rather depends on a basic implementation of the
reader. These tests thus run fast and are cheap to maintain (they don't
depend on the implementation but on the interface).

The fake reader is implemented as an in-memory list of authorizations
that can be created so they can be accessed using getByKey and search.
The search is implemented as simple as possible, only considering filter
options that are not null. The fake is closed after all tests (clearing
the stored authorizations), which is not strictly necessary. However, I
felt this highlighted how this data store is available throughout all
tests.

Note that an alternative way to test the AuthorizationChecker would have
been to set up Elasticsearch using Testcontainers. However, I found this
to be slow and brittle. The current tests using the fake reader run in a
fraction of a second instead.

korthout committed:
test: cover collectPermissionTypes method

Adds another ParameterizedTest that covers the collectPermissionTypes
method of AuthorizationChecker. Again we use FakeAuthorizationReader to
easily set up the available authorization entities. But different from
the IsAuthorizedTests, here we set up a new FakeAuthorizationReader for
each test. This allows us to easily set up different authorizations for
each scenario.

The tests also use a dedicated CollectPermissionScenario which offers a
builder pattern to clarify the scenarios in more details. The build
method decides whether this concerns anonymous access or authenticated.

Some of the scenarios overlap in what edge case they cover. For example,
there are several mismatch scenarios which cover the case where the
resource id does not match any of the available authorizations. Another
example are the scenarios where a union of permissions is formed across
different authorizations. Still, I found it useful to add these as to
cover a wider range of different AuthorizationResourceTypes.
Additionally, these tests are cheap to maintain and run fast so there's
no real cost to having too many scenarios.

The authEntity method helps easily create AuthorizationEntity objects
where each gets a unique key.

korthout committed:
refactor: remove unnecessary mockito usage

Mockito is slow, and leads to tests depending on the implementation
rather than on the interface, making it costly to maintain.

korthout committed:
test: cover retrieveAuthorizedAuthorizationScopes

Adds a similar test as was provided for the other methods focussing on
the retrieveAuthorizedAuthorizationScopes. As this uses the same methods
under the hood, we do not expect anything new here. We're just adding
some coverage of different access patterns, given some different
authorization scenarios.

Again, some of the test scenarios overlap, but that's fine as these
tests are cheap, running fast and not depending on the implementation.

Like the other tests we again use a dedicated Scenario with builder
pattern to clarify the scenarios in detail.

korthout committed:
test: add mapping rule matcher tests

This adds some basic unit tests to cover different edge cases of the
MappingRuleMatcher. It's set up as a nested test as we only test the
matchingRules method for now. If the class is expanded in the future, we
can easily support the testing of them as well.

The tests are quite straightforward otherwise, covering some basic
JSONPath concepts like nested paths, as well as error cases like invalid
expressions, and non-matching types.

Lastly, I've added a test case to cover the internal caching. It does
not actually test that the caching works as expected, but just to ensure
that some of its implementation is covered specifically in these tests.

ana-vinogradova-camunda reviewed (approved):
LGTM! 🚀 
Thank you, @korthout !

korthout committed:
refactor: clarify assertion expectation

The conditional was a bit difficult to read. It can be simplified with a
simple enum method.

korthout committed:
refactor: align imports to styleguide

ana-vinogradova-camunda commented on security/security-services/src/test/java/io/camunda/security/impl/AuthorizationCheckerTest.java:
```diff
@@ -76,4 +93,347 @@ public void notAuthorizedWhenOwnerIdsIsEmpty() {
+      // then
+      Assertions.assertThat(isAuthorized)
+          .describedAs(displayName)
+          .isEqualTo(expected == Expected.ACCESS_ALLOWED);
```
🚀 thank you @korthout
